### PR TITLE
Enables communication between the webview and RN

### PR DESCRIPTION
**Pull request by odino**
Android devs have been relying for ages on intercepting custom HTTP
calls made by the webview, to map them to native actions,
[like here](https://www.caphal.com/android/communication-between-application-and-webview/).

Say that we have a "close" button in our web interface. We first detect
if we're running on web or within a webview -- if on web, we simply call `window.close()`.
If we're running within the app things get a bit more complicated, but we can
simply call `app://exit` and the app will be able to intercept this call and
close the webview for us.

Now, this PR adds the ability of doing it on RN. You basically specify a
protocol for messages that should be forwarded to RN (ie. `rn`) and then,
when your webview loads a resource from `rn://xyz` the `onMessage` callback
will be called. You can access the URL etc etc through `event.nativeEvent.url`.

Example:

``` javascript
let onMessage = (event) => {
  let {url} = event.nativeEvent;

  if (url.endsWith('exit')) {
    BackAndroid.exitApp(); // just sayin'...
  }
}

return <Webview
  source={...}
  messageProtocol={'myApp'}
  onMessage={onMessage}
>
```

Or something of that sort. Communicating between the webview and RN / the ability
to intercept resources loaded within the webview has been already discussed but I
feel we need a very simple implementation rather than creating a full-blown
messaging pattern which will just hurt (maintainability, complexity) in the long run.

Also, my Java is pretty rusty ;-)

### CODE REVIEW

**Comment by odino** [comment]
My biggest doubt here is that this guy is android-specific. I think we already have something working on [IOS](https://facebook.github.io/react-native/docs/webview.html#onshouldstartloadwithrequest) so I would prefer to see both working in the same way. Anyhow, let me know what you guys think.

**Comment by odino** [comment]
Saw @astreet comments on https://github.com/facebook/react-native/pull/6478#issuecomment-208988259 and I agree that regex patterns might be a good, viable solution. Though, what if we need to run cb when a pattern is matched?

File: ReactAndroid/src/main/java/com/facebook/react/views/webview/events/MessageEvent.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package com.facebook.react.views.webview.events;
+
+import com.facebook.react.bridge.WritableMap;
+import com.facebook.react.uimanager.events.Event;
+import com.facebook.react.uimanager.events.RCTEventEmitter;
+
+/**
+ * Event emitted when there is an error in loading.
```
Comment (mkonicek) [inline ReactAndroid/src/main/java/com/facebook/react/views/webview/events/MessageEvent.java]: Is this Javadoc correct?

**Comment by mkonicek** [comment]
In the example, you have: `messageProtocol={'myApp'}`. Does the WebView load a `myApp://` URL? Can you provide a full example please?

**Comment by mkonicek** [comment]
> Saw @astreet comments on #6478 (comment) and I agree that regex patterns might be a good, viable solution. Though, what if we need to run cb when a pattern is matched?

Can you elaborate please? What do you mean by "what if we need to run cb when a pattern is matched?"?

**Comment by odino** [comment]
This is what I meant -- we could pass some object like this:

```
let routesToIntercept = {
  'myApp://some-action': () => console.log(123),
  'myApp://*': () => console.log(456)
}

return <Webview
  routesToIntercept={routes}
/>
```

The problem I see with my PR is that IOS / android implementations will vary quite a bit, which sucks. So Im open to consider changing the implementation in favor of something that can work longer term. Let me know your thoughts.

File: ReactAndroid/src/main/java/com/facebook/react/views/webview/events/MessageEvent.java (hunk 1 above)
Comment (odino) [inline ReactAndroid/src/main/java/com/facebook/react/views/webview/events/MessageEvent.java]: nope, sorry -- copypasta :)

**Comment by stereodenis** [comment]
@odino https://github.com/alinz/react-native-webview-bridge

**Comment by odino** [comment]
@stereodenis that wont work for legacy implementations that are based on calling custom URLs :) even though I think it's a good solution I was aiming at taking care of the practice of using custom URLs since there's a lot of android + webview implementations that use that method.

**Comment by facebook-github-bot-3** [comment]
@mkonicek would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by stereodenis** [comment]
@odino will it be replacement for react-native-webview-bridge?

**Comment by odino** [comment]
@stereodenis well, I would like to see a default way to implement some sort of communication between webviews and RN :)

**Comment by stereodenis** [comment]
https://github.com/alinz/react-native-webview-bridge/issues/109

**Comment by mkonicek** [comment]
Sorry for the delay on reviewing this - there's now a related proposal using `postMessage`: https://github.com/facebook/react-native/pull/9762

**Comment by lacker** [comment]
Since the `postMessage` change is accepted with https://github.com/facebook/react-native/pull/9762 I think this pull request has been superceded. I'm going to close this PR but please feel free to reopen if this is hitting a different use case or has an independent rationale.

**Pull request closed without merging**